### ISSUE TITLE
Is this really a typo?

### ISSUE THREAD

**Issue by rmcgibbo** [jtojnar/nixpkgs-hammering#33]
https://github.com/NixOS/nixpkgs/pull/112817#issuecomment-777802684

**Comment by jtojnar** [comment]
Hmm, yeah, the [known attributes list](https://github.com/jtojnar/nixpkgs-hammering/blob/master/lib/derivation-attributes.nix) is still incomplete. I wonder where does `phases` fit in the attribute ordering. cc @worldofpeace

Or perhaps we should just add extra known attributes to the [`attribute-typo` check](https://github.com/jtojnar/nixpkgs-hammering/blob/00a60d6010e90888ada900569e215c5c481f3b09/overlays/attribute-typo.nix#L12) and delay thinking about it for now.

**Comment by rmcgibbo** [comment]
I wonder if we could just screen all of nixpkgs today and then filter the list of typos that it identifies by hand into true positives and false positives?

**Comment by worldofpeace** [comment]
> Hmm, yeah, the [known attributes list](https://github.com/jtojnar/nixpkgs-hammering/blob/master/lib/derivation-attributes.nix) is still incomplete. I wonder where does `phases` fit in the attribute ordering. cc @worldofpeace
> 
> Or perhaps we should just add extra known attributes to the [`attribute-typo` check](https://github.com/jtojnar/nixpkgs-hammering/blob/00a60d6010e90888ada900569e215c5c481f3b09/overlays/attribute-typo.nix#L12) and delay thinking about it for now.

So I would say `phases` is could be the initialization/definition of the phases we're going to see defined.
These could be named anything... So I'd think it would be
```
phases = [ "somePhase"];

somePhase = ''
'';

so it would essentially replace the phase ordering, and be skipped because there's no convention as to what those should be called. Unless we want a convention defined like camelcasing for them.

**Comment by jtojnar** [comment]
@rmcgibbo How did you do the mass test on all python packages? It would be useful to get the list of unknown attributes with the following patch:

```patch
diff --git a/overlays/attribute-typo.nix b/overlays/attribute-typo.nix
index 251124a..bd2fe91 100644
--- a/overlays/attribute-typo.nix
+++ b/overlays/attribute-typo.nix
@@ -69,7 +69,7 @@ let
           in {
             name = "attribute-typo";
             # Comment out the following line to see all unknown attributes (when extending our lexicon).
-            cond = builtins.length suggestions > 0;
+            cond = true;
             msg = ''
               A likely typo in `${argName}` argument was found${prettySuggestions suggestions}
             '';
diff --git a/tools/nixpkgs-hammer b/tools/nixpkgs-hammer
index a046285..b1166be 100755
--- a/tools/nixpkgs-hammer
+++ b/tools/nixpkgs-hammer
@@ -174,7 +174,7 @@ def main(args):
     # We do it by using functions that return overlays so we need to instantiate them.
     overlay_expressions = []
     for overlay_generator in overlay_generators_path.glob('*'):
-        if overlay_generator.stem in args.excluded_rules:
+        if overlay_generator.stem != 'attribute-typo':
             continue
 
         overlay = overlay_generator.name
@@ -209,7 +209,7 @@ def main(args):
         print('Nix expression:', all_messages_nix, file=sys.stderr)
 
     all_overlay_messages = nix_eval_json(all_messages_nix, args.show_trace)
-    all_ast_check_messages = ast_checks(args.nix_file, args.attr_paths, args.excluded_rules)
+    all_ast_check_messages = {}
     all_messages = concatenate_messages(all_overlay_messages, all_ast_check_messages)
 
     if args.json:
```

**Comment by rmcgibbo** [comment]
> @rmcgibbo How did you do the mass test on all python packages? 

For the ast-check against all of the python packages, I basically just did `cargo run --bin <check-name> ~/path/to/nixpkgs/pkgs/development/python-modules/*/default.nix`, but I can do something different for this one since it's overlay-based. Give me a few minutes.

**Comment by jtojnar** [comment]
For now, I found https://github.com/NixOS/nixpkgs/search?q=attribute-typo+is%3Apr&type=issues. Looks like levenshtein distance 4 is really too big, especially for 6 letter words (on https://github.com/NixOS/nixpkgs/pull/112821#issuecomment-777840588, it was suggested `PROTOC` → `src`).

**Comment by rmcgibbo** [comment]
This is taking longer to run that I expected, so it's not yet a complete list, but here: https://gist.github.com/rmcgibbo/6b6630362aa6c7409035d654c2eda4a5

**Comment by rmcgibbo** [comment]
I doubt it will uncover much more that's particularly useful since there's a long tail of typos, but the evaluation is still running and will probably run for another 1hr or so.

**Comment by rmcgibbo** [comment]
I updated the gist above with the final results of my sweep.

**Comment by rmcgibbo** [comment]
Closed by #38
